### PR TITLE
feat: Adds the ability to search emojis and render

### PR DESCRIPTION
**Pull request by iamsivin**
# Pull Request Template

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
https://www.loom.com/share/a2798dd9fbd44320a495353dbc798704

**Test cases**
1. Search emojis from the emoji box search.
2. See all places where the emoji box Is used. (widget, reply box, expanded reply box)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/javascript/shared/components/emoji/EmojiInput.vue [hunk 1]
```diff
@@ -44,14 +94,65 @@ export default {
...
     return {
-      selectedKey: 'Smileys & Emotion',
+      selectedKey: 'Search',
       emojis,
+      search: '',
     };
   },
+  computed: {
+    categories() {
+      return this.emojis.map(category => category);
```
Comment (nithindavid) [inline app/javascript/shared/components/emoji/EmojiInput.vue:104]:
```suggestion
      return [...this.emojis]
```

File: app/javascript/shared/components/emoji/EmojiInput.vue [hunk 2]
```diff
@@ -44,14 +94,65 @@ export default {
...
+        const allEmojis = category.emojis.filter(emoji =>
+          emoji.name.includes(this.search)
+        );
+        return allEmojis.length > 0
+          ? { ...category, emojis: allEmojis }
+          : { ...category, emojis: [] };
+      });
+    },
+    hasNoSearch() {
+      return this.selectedKey !== 'Search' && this.search === '';
```
Comment (nithindavid) [inline app/javascript/shared/components/emoji/EmojiInput.vue:123]: Can we make a constant for search key `SEARCH_KEY`

**Review by nithindavid — CHANGES_REQUESTED** [review]
MInor changes added.
Great PR 🔥

File: app/javascript/shared/components/emoji/emojisGroup.json [hunk 3]
```diff
@@ -0,0 +1,9296 @@
+[
```
Comment (pranavrajs) [inline app/javascript/shared/components/emoji/emojisGroup.json:1]: @iamsivin Can you please share the reference link from where you have taken this? 

We can update it to v15.0 (https://unicode.org/emoji/charts/emoji-list.html) if you have not done that already.

**Review by pranavrajs — CHANGES_REQUESTED** [review]
@iamsivin Nice work! Please see the feedback here. https://www.loom.com/share/55f4004c850940d4a792a1995a332d20

1. As mentioned above if we can conform to v15.0 of emojis, that would be nice. (not mandatory though)
2. Search does not return results sometimes.
3. The arrows were removed in this update. I felt that the emoji picker is better with the arrows. 
Screenshot 2022-11-24 at 11 15 38 AM
4. The changes introduce a large payload so please add dynamic imports to the places where the emoji picker is used.

File: app/javascript/shared/components/emoji/emojisGroup.json (hunk 3 above)
Comment (iamsivin) [reply app/javascript/shared/components/emoji/emojisGroup.json:1]: Taken it from [here](https://github.com/muan/unicode-emoji-json/blob/main/data-by-group.json) and edited the payload as we needed, and checked it with v15.0 (https://unicode.org/emoji/charts/emoji-list.html)

File: app/javascript/shared/components/emoji/EmojiInput.vue (hunk 2 above)
Comment (iamsivin) [reply app/javascript/shared/components/emoji/EmojiInput.vue]: Done

File: app/javascript/shared/components/emoji/EmojiInput.vue (hunk 1 above)
Comment (iamsivin) [reply app/javascript/shared/components/emoji/EmojiInput.vue]: Done

File: app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue [hunk 4]
```diff
@@ -964,13 +965,13 @@ export default {
...
-  bottom: 12px;
+  bottom: var(--space-slab);
   left: -320px;
   right: unset;
 
   &::before {
-    right: -16px;
-    bottom: 10px;
+    right: var(--space-minus-normal);
+    bottom: var(--space-one);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue:974]: Please stick to `4, 8, 16` for spacing.

File: app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue [hunk 5]
```diff
@@ -964,13 +965,13 @@ export default {
 
 .emoji-dialog {
   top: unset;
-  bottom: 12px;
+  bottom: var(--space-slab);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue:968]: Please stick to `4, 8, 16` for spacing.

File: app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue [hunk 6]
```diff
@@ -983,8 +984,8 @@ export default {
 
   &::before {
     transform: rotate(0deg);
-    left: var(--space-smaller);
-    bottom: var(--space-minus-slab);
+    left: var(--space-half);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue:987]: Please stick to `4, 8, 16` for spacing.

File: app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue [hunk 7]
```diff
@@ -983,8 +984,8 @@ export default {
...
+    bottom: var(--space-minus-one);
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ReplyBox.vue:988]: Please stick to `-4, -8, -16` for spacing.

File: app/javascript/shared/components/emoji/EmojiInput.vue [hunk 8]
```diff
@@ -93,59 +199,104 @@ $color-bg: #ebf0f5;
...
+.emoji-search--wrap {
+  background-color: $color-white;
+  margin: $space-small;
+  position: sticky;
+  top: $space-small;
+
+  .emoji-search--input {
+    background-color: $color-bg;
+    border: 1px solid transparent;
+    border-radius: 5px;
```
Comment (nithindavid) [inline app/javascript/shared/components/emoji/EmojiInput.vue:243]: Use variable here

File: app/javascript/shared/components/emoji/EmojiInput.vue [hunk 9]
```diff
@@ -93,59 +199,104 @@ $color-bg: #ebf0f5;
...
+  border: 0;
+  border-radius: $space-smaller;
+  cursor: pointer;
+  font-size: $font-size-medium;
+  height: $space-medium;
+  margin: 0;
+  padding: 0 $space-smaller;
+
+  &:hover {
+    background: $color-bg;
```
Comment (nithindavid) [inline app/javascript/shared/components/emoji/EmojiInput.vue:220]: Css variables for colors are available for both the widget and dashboard.

File: app/javascript/shared/components/emoji/EmojiInput.vue [hunk 10]
```diff
@@ -93,59 +199,104 @@ $color-bg: #ebf0f5;
...
+  flex-direction: column;
+  height: 212px;
+  justify-content: center;
 
-    .emoji--item {
-      margin: $space-smaller;
-      line-height: 1.5;
-    }
+  .emoji-icon {
+    color: $color-text-light;
```
Comment (nithindavid) [inline app/javascript/shared/components/emoji/EmojiInput.vue:264]: Css variables for colors are available for both the widget and dashboard. for other occurrences as well

**Review by nithindavid — CHANGES_REQUESTED** [review]
Please see the comments

**Comment by iamsivin** [comment]
> @iamsivin Found some bugs while testing this. Please see the video - https://www.loom.com/share/8d49f63085bc4922a5feaf45763831dd

@nithindavid Done https://github.com/chatwoot/chatwoot/pull/5928/commits/4369bdfe01b3b7f0ad39441df1d7f9e546614fb2

**Comment by nithindavid** [comment]
Merging this with a minor bug found in the normal editor. Can be be fixed later - https://github.com/chatwoot/product/issues/697

**Review by nithindavid — APPROVED** [review]

**Review by fayazara — APPROVED** [review]
Working as expected. 💯 

Although can we close the emoji picker once we select an emoji, since most of the native emoji picker works this way, noticed this when I used the fill width layout and it was covering my ReplyBox. Thoughts? @nithindavid @pranavrajs

**Review by muhsin-k — APPROVED** [review]

**Comment by pranavrajs** [comment]
>  Although can we close the emoji picker once we select an emoji, since most of the native emoji picker works this way, noticed this when I used the fill width layout and it was covering my ReplyBox. Thoughts? @nithindavid @pranavrajs

@fayazara Let us pick it up as an enhancement.

**Pull request merged**